File: helper/launcher/csharp/Logger.cs
using System;
using System.IO;

namespace ShippingManagerCoPilot.Launcher
{
    public static class Logger
    {
        private static readonly object _lock = new();
        private static string? _logFilePath;
        private static bool _initialized = false;

        private static string LogFilePath
        {
            get
            {
                if (_logFilePath == null)
                {
                    var logsDir = Path.Combine(App.UserDataDirectory, "logs");
                    Directory.CreateDirectory(logsDir);
                    _logFilePath = Path.Combine(logsDir, "launcher.log");
                }
                return _logFilePath;
            }
        }

        /// <summary>
        /// Initialize logger - clears the log file on first call
        /// </summary>
        private static void EnsureInitialized()
        {
            if (_initialized) return;
            _initialized = true;

            try
            {
                // Clear the log file on startup
                File.WriteAllText(LogFilePath, string.Empty);
            }
            catch
            {
                // Ignore errors during initialization
            }
        }

        public static void Log(string level, string message)
        {
            var timestamp = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss");
            var line = $"[{timestamp}] [{level.ToUpper()}] [Launcher] {message}";

            lock (_lock)
            {
                try
                {
                    EnsureInitialized();
                    File.AppendAllText(LogFilePath, line + Environment.NewLine);
                }
                catch
                {
                    // Ignore logging errors
                }
            }

            // Also write to debug output
            System.Diagnostics.Debug.WriteLine(line);
        }

        public static void Info(string message) => Log("INFO", message);
        public static void Debug(string message) => Log("DEBUG", message);
        public static void Warn(string message) => Log("WARN", message);
        public static void Error(string message) => Log("ERROR", message);
    }
}

File: helper/launcher/nodejs/logger.js
/**
 * @fileoverview Launcher Logger
 * @module launcher/logger
 *
 * Writes logs to both console and server.log for consistency with server logging.
 */

const fs = require('fs');
const path = require('path');

let logFile = null;
let logStream = null;

/**
 * Detect if running as packaged SEA executable
 * @returns {boolean}
 */
function isPackaged() {
  try {
    const sea = require('node:sea');
    return sea.isSea();
  } catch {
    return false;
  }
}

/**
 * Initialize log file - writes to server.log in userdata/logs
 */
function initLogFile() {
  if (logStream) return;

  try {
    // Determine log directory
    let logsDir;

    if (isPackaged()) {
      const appDataDir = process.env.LOCALAPPDATA || process.env.HOME;
      logsDir = path.join(appDataDir, 'ShippingManagerCoPilot', 'userdata', 'logs');
    } else {
      // 3 levels up from helper/launcher/nodejs to project root
      logsDir = path.join(__dirname, '..', '..', '..', 'userdata', 'logs');
    }

    // Ensure directory exists
    if (!fs.existsSync(logsDir)) {
      fs.mkdirSync(logsDir, { recursive: true });
    }

    logFile = path.join(logsDir, 'server.log');
    logStream = fs.createWriteStream(logFile, { flags: 'a' });
  } catch (err) {
    // Silently fail - logging shouldn't crash the app
    console.error('Failed to init log file:', err.message);
  }
}

/**
 * Log message with timestamp
 * @param {string} level - Log level (info, warn, error, debug)
 * @param {string} message - Message
 */
function log(level, message) {
  const now = new Date();
  const timestamp = now.toLocaleString('sv-SE', { timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone }).replace(' ', 'T');
  const line = `[${timestamp}] [${level.toUpperCase()}] [Startup] ${message}`;

  // Always try console
  console.log(line);

  // Write to file
  if (!logStream) {
    initLogFile();
  }

  if (logStream) {
    logStream.write(line + '\n');
  }
}

function info(message) {
  log('info', message);
}

function warn(message) {
  log('warn', message);
}

function error(message) {
  log('error', message);
}

function debug(message) {
  log('debug', message);
}

module.exports = {
  log,
  info,
  warn,
  error,
  debug
};
